onnxruntime/core/optimizer/unsqueeze_elimination.cc

```diff
@@ -2,104 +2,108 @@
+  std::vector<int64_t> axes;
+  for (int i = 0; i < attr->ints_size(); i++) {
+    axes.push_back(static_cast<int64_t>(attr->ints(i)));
+  }
 
-    const onnxruntime::NodeAttributes& attributes = node.GetAttributes();
-    const ONNX_NAMESPACE::AttributeProto* attr = &attributes.find("axes")->second;
-    if (attr == nullptr || attr->type() != AttributeProto_AttributeType_INTS) {
-      continue;
-    }
+  // Generate new dims
+  NodeArg* input_def = node.MutableInputDefs()[0];
+  const ONNX_NAMESPACE::TensorProto* tensor_proto = nullptr;
+  graph.GetInitializedTensor(input_def->Name(), tensor_proto);
+  if (tensor_proto == nullptr) {
+    return Status::OK();
+  }
+  std::vector<int64_t> new_dims(axes.size() + tensor_proto->dims().size(), 0);
+  if (new_dims.size() >= std::numeric_limits<int>::max()) {
+    return Status(ONNXRUNTIME, FAIL, "index out of range");
+  }
 
-    // Get attribute of "axes"
-    std::vector<int64_t> axes;
-    for (int i = 0; i < attr->ints_size(); i++) {
-      axes.push_back(static_cast<int64_t>(attr->ints(i)));
-    }
+  for (int64_t axis : axes) {
+    new_dims[axis] = 1;
+  }
 
-    // Generate new dims
-    NodeArg* input_def = node.MutableInputDefs()[0];
-    const ONNX_NAMESPACE::TensorProto* tensor_proto = nullptr;
-    graph.GetInitializedTensor(input_def->Name(), tensor_proto);
-    if (tensor_proto == nullptr) {
-      continue;
+  auto begin = tensor_proto->dims().cbegin();
+  for (auto& axis : new_dims) {
+    if (axis == 0) {
+      axis = *begin++;
     }
-    std::vector<int64_t> new_dims(axes.size() + tensor_proto->dims().size(), 0);
-    if (new_dims.size() >= std::numeric_limits<int>::max())
-      return Status(ONNXRUNTIME, FAIL, "index out of range");
+  }
 
-    for (int64_t axis : axes) {
-      new_dims[axis] = 1;
-    }
+  // Update shape of tensor proto
+  ONNX_NAMESPACE::TensorProto new_tensor_proto(*tensor_proto);
 
-    auto begin = tensor_proto->dims().cbegin();
-    for (auto& axis : new_dims) {
-      if (axis == 0) {
-        axis = *begin++;
-      }
+  for (int i = 0; i < static_cast<int>(new_dims.size()); i++) {
+    if (i < tensor_proto->dims().size()) {
+      new_tensor_proto.set_dims(i, new_dims[i]);
+    } else {
+      new_tensor_proto.add_dims(new_dims[i]);
     }
+  }
+  graph.RemoveInitializedTensor(input_def->Name());
+  graph.AddInitializedTensor(new_tensor_proto);
 
-    // Update shape of tensor proto
-    ONNX_NAMESPACE::TensorProto new_tensor_proto(*tensor_proto);
+  // Update shape of NodeArg
+  TensorShapeProto shape;
+  for (auto dim : new_dims) {
+    shape.add_dim()->set_dim_value(dim);
+  }
+  input_def->SetShape(shape);
+
+  // Keep all output nodes in a list (after removing all edges below, there will be no other way to access them; 
+  // I cannot first update the output defs, because edge removal will not work afterwards).
+  std::vector<NodeIndex*> output_nodes_idx;
+  for (auto it = node.OutputNodesBegin(), end = node.OutputNodesEnd(); it != end; ++it) {
```

Comment:
nit: as NodeIndex is size_t you don't need to use a NodeIndex*

Reply:
Will do, although I will probably simplify this part by a change in the Graph::RemoveEdge.